crates/holochain/src/conductor/cell/authority.rs

```diff
@@ -167,37 +169,88 @@ pub async fn handle_get_entry(
+        .get_activity_status(&agent)?
+        .unwrap_or(ChainStatus::Empty);
+    let highest_observed = meta_integrated.get_activity_observed(&agent)?;
 
-    if options.include_activity {
+    // TODO: If full headers aren't requested then query doesn't work
+
+    // Valid headers
+    let valid_activity = if options.include_valid_activity {
         fresh_reader!(env, |r| {
-            let activity: Vec<_> = meta_integrated
-                .get_activity(&r, ChainItemKey::Agent(agent.clone()))?
-                .filter_map(|h| element_integrated.get_header(&h.header_hash))
-                .filter(|shh| Ok(query.check(shh.header())))
+            let hashes = meta_integrated
+                .get_activity_sequence(&r, ChainItemKey::Agent(agent.clone()))?
                 .collect()?;
-            // TODO: Return the chain status and check rejected
-            Ok(AgentActivity::valid(activity))
-        })
+            if options.include_full_headers {
+                CellResult::Ok(Activity::Full(get_full_headers(
+                    hashes,
+                    element_integrated,
+                    &r,
+                )?))
+            } else {
+                Ok(Activity::Hashes(hashes))
+            }
+        })?
     } else {
+        Activity::NotRequested
+    };
+
+    // Rejected hashes
+    let rejected_activity = if options.include_rejected_activity {
         fresh_reader!(env, |r| {
-            match meta_integrated
-                .get_activity(&r, ChainItemKey::Agent(agent))?
-                .last()?
-            {
-                Some(h) => {
-                    // TODO: Just get the sequence number from the key (Doable in the next PR)
-                    match element_integrated.get_header(&h.header_hash)? {
-                        Some(shh) => Ok(AgentActivity::valid_without_activity(shh.header())),
-                        None => Ok(AgentActivity::empty()),
-                    }
-                }
-                None => Ok(AgentActivity::empty()),
+            let hashes = meta_rejected
+                .get_activity_sequence(&r, ChainItemKey::Agent(agent.clone()))?
+                .collect()?;
+            if options.include_full_headers {
+                CellResult::Ok(Activity::Full(get_full_headers(
+                    hashes,
+                    element_rejected,
+                    &r,
+                )?))
+            } else {
+                Ok(Activity::Hashes(hashes))
             }
-        })
-    }
+        })?
+    } else {
+        Activity::NotRequested
+    };
+
+    Ok(AgentActivity {
+        valid_activity,
+        rejected_activity,
+        agent,
+        status,
+        highest_observed,
+    })
+}
+
+fn get_full_headers<P: PrefixType, R: Readable>(
+    hashes: Vec<(u32, HeaderHash)>,
+    database: ElementBuf<P>,
+    reader: &R,
+) -> CellResult<Vec<SignedHeaderHashed>> {
+    info!(line = line!());
+    let now = std::time::Instant::now();
```

Comment:
should we do some kind of `#[cfg(debug_assertions)]` thing so we're not querying the system clock on release builds?

Reply:
This is where the current bug is showing up D-01161. I'm kind of keen to leave this here for now as it will help in debugging but will definitely remove it once the bug is solved